### PR TITLE
 Add support to match_phrase query for zero_terms_query.

### PR DESCRIPTION
**Pull request by jtibshirani**
Addresses #29344.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java [hunk 1]
```diff
@@ -39,6 +40,7 @@
 public class MatchPhraseQueryBuilder extends AbstractQueryBuilder<MatchPhraseQueryBuilder> {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java]: The additions in this class are a bit messy, but I favored consistency with `MatchQueryBuilder` and `MultiMatchQueryBuilder`.

File: server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java [hunk 2]
```diff
@@ -67,6 +71,7 @@ public MatchPhraseQueryBuilder(StreamInput in) throws IOException {
         fieldName = in.readString();
         value = in.readGenericValue();
         slop = in.readVInt();
+        zeroTermsQuery = ZeroTermsQuery.readFromStream(in);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java]: You'll need to handle backward compatibility here. You can check https://github.com/elastic/elasticsearch/blob/5263b8cc7ee7b1a52880ab98587b88d6c00557b8/server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java#L132 for an example (the version should be v7_0...).

File: server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java [hunk 3]
```diff
@@ -75,6 +80,7 @@ protected void doWriteTo(StreamOutput out) throws IOException {
         out.writeString(fieldName);
         out.writeGenericValue(value);
         out.writeVInt(slop);
+        zeroTermsQuery.writeTo(out);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java]: Same here, write only the `out.getVersion` is before 7.

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @jtibshirani 
It looks great, I left one comment regarding backward compatibility that needs to handle in the (de)serialization of the query.

**Comment by jtibshirani** [comment]
Fixed, thanks!

File: server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java (hunk 3 above)
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java]: Guessing you mean 'after 7' here.

**Review by jimczi — APPROVED** [review]
LGTM
You can also backport to 6x. Don t need to open a new pr though, you can just change the version for bwc when you merge in the 6x branch. Then after you commmited in 6x you can change the version in master to reflect the version compatibility.

File: server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java (hunk 3 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java]: Yep, sorry

**Pull request merged**